Compiler/src/translate.ml

```diff
@@ -0,0 +1,1554 @@
+    let inouts =  (arg_regs @ [return_temp]) in
+    let inst_op : Program_types.instruction_operation = if is_spread then 
+        (let op : Operations_types.create_array_with_spread = Operations_types.{spreads = is_spread_list} in
+            Program_types.Create_array_with_spread op)
+        else Program_types.Create_array in
+    let inst : Program_types.instruction = Program_types.{
+        inouts = inouts;
+        operation = inst_op;
+    } in
+    (return_temp, flat_inst @ [inst])
+
+
+and proc_create_object_property (prop_val: ('M, 'T) Flow_ast.Expression.Object.property) (tracker: Context.tracker) =
+    match prop_val with
+        Property (_, prop) ->
+            let temp_reg, prop_name_key, inst = match prop with
+                Init init_val ->
+                    let temp, exp_inst = proc_expression init_val.value tracker in
+                    (temp, init_val.key, exp_inst)
+                | Set func -> 
+                    let (_, act_func) = func.value in
+                    let temp, inst = proc_func act_func tracker false in
+                    (temp, func.key, inst)
+                | Get func -> 
+                    let (_, act_func) = func.value in
+                    let temp, inst = proc_func act_func tracker false in
+                    (temp, func.key, inst)
+                | Method func -> 
+                    let (_, act_func) = func.value in
+                    let temp, inst = proc_func act_func tracker false in
+                    (temp, func.key, inst) in
+            let prop_name : string = match prop_name_key with
+                Literal (_, l) -> l.raw
+                | Identifier (_, i) -> i.name
+                | PrivateName (_, p) -> let (_, i) = p.id in
+                    i.name
+                | Computed _ -> raise (Invalid_argument "Unhandled Object key type Computed Key in object creation") in
+            ((temp_reg, [prop_name]), inst)
+        | SpreadProperty (_, spreadProp) -> 
+            let temp_reg, exp_inst = proc_expression spreadProp.argument tracker in
+            ((temp_reg, []), exp_inst)
+
+and proc_create_object (exp : ('M, 'T) Flow_ast.Expression.Object.t) (tracker: Context.tracker) =
+    let props = exp.properties in
+    let temp_func a = proc_create_object_property a tracker in
+    let (obj_temp_tuple, create_obj_inst) = List.split (List.map temp_func props) in
+    let (obj_temp_list, obj_key_list_unflattened) = List.split obj_temp_tuple in
+    let obj_key_list_flat = List.flatten obj_key_list_unflattened in
+    let flat_inst = List.flatten create_obj_inst in
+    let return_temp = Context.get_new_intermed_temp tracker in
+    let inouts =  (obj_temp_list @ [return_temp]) in
+    let op = if List.length obj_key_list_flat == List.length obj_temp_list then
+            Program_types.Create_object{property_names = obj_key_list_flat} 
+        else
+            Program_types.Create_object_with_spread{property_names = obj_key_list_flat} 
```

Comment:
I think I'll refactor the CreateObject and CreateArray operations with the upcoming type system refactoring so that they
- Contain the resulting output type (so that the AbstractInterpreter can be smarter)
- Support Get, Set, and Method syntax (for `CreateObject`) which automatically expose `this`
- Spreading (not more `WithSpread` variants)
- Holes (in the case of arrays)

That should hopefully make this logic a bit simpler. WDYT? Anything else that's missing?